jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ToolkitAuthManager.kt

```diff
@@ -204,23 +205,42 @@ fun AwsBearerTokenConnection.lazyIsUnauthedBearerConnection(): Boolean {
     return false
 }
 
-fun reauthProviderIfNeeded(connection: ToolkitConnection): BearerTokenProvider {
+fun reauthProviderIfNeeded(project: Project?, connection: ToolkitConnection): BearerTokenProvider {
     val tokenProvider = (connection.getConnectionSettings() as TokenConnectionSettings).tokenProvider.delegate as BearerTokenProvider
+
+    return reauthProviderIfNeeded(project, tokenProvider)
+}
+
+fun reauthProviderIfNeeded(project: Project?, tokenProvider: BearerTokenProvider): BearerTokenProvider {
+    maybeReauthProviderIfNeeded(project, tokenProvider) {
+        runUnderProgressIfNeeded(project, message("credentials.sono.login.pending"), true) {
+            tokenProvider.reauthenticate()
+        }
+    }
+
+    return tokenProvider
+}
+
+fun maybeReauthProviderIfNeeded(project: Project?, tokenProvider: BearerTokenProvider, onReauthRequired: (SsoOidcException?) -> Any) {
     val state = tokenProvider.state()
-    runUnderProgressIfNeeded(null, message("settings.states.validating.short"), true) {
-        if (state == BearerTokenAuthState.NEEDS_REFRESH) {
+    when (state) {
+        BearerTokenAuthState.NOT_AUTHENTICATED -> {
+            getLogger<ToolkitAuthManager>().info { "Token provider NOT_AUTHENTICATED, requesting login" }
+            onReauthRequired(null)
+        }
+
+        BearerTokenAuthState.NEEDS_REFRESH -> {
             try {
-                tokenProvider.resolveToken()
-                BearerTokenProviderListener.notifyCredUpdate(tokenProvider.id)
+                runUnderProgressIfNeeded(project, message("credentials.sono.login.refreshing"), true) {
+                    tokenProvider.resolveToken()
+                    BearerTokenProviderListener.notifyCredUpdate(tokenProvider.id)
+                }
             } catch (e: SsoOidcException) {
-                tokenProvider.reauthenticate()
+                getLogger<ToolkitAuthManager>().warn(e) { "Redriving AWS Builder ID login flow since token could not be refreshed" }
+                onReauthRequired(e)
```

Comment:
Do we need to use the passed-in parameter `e`? I don't see it's currently being used.

Reply:
optional